### PR TITLE
Add support for write barrier in dynamic noc mode

### PR DESCRIPTION
**Pull request by yugaoTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/15339

### Problem description
previously, the dynamic noc will hang when user calls barrier, this is due to the status register accounts for both RISCs transaction (ie, BRISC and NCRISC both increments the status reg when issuing on the same noc).
with this change, we allocate a scratch memory for WH and GS, and use it as software counters. For BH we have extra scratch register so no need for the scratch memory.

### Checklist
- [x] Post commit CI https://github.com/tenstorrent/tt-metal/actions/runs/12126931245
- [x] Blackhole Post  https://github.com/tenstorrent/tt-metal/actions/runs/12127036053
- [x] Model regression CI https://github.com/tenstorrent/tt-metal/actions/runs/12129141464
- [x] Device performance regression CI  https://github.com/tenstorrent/tt-metal/actions/runs/12129148380
- [x] t3k frequent https://github.com/tenstorrent/tt-metal/actions/runs/12131084677
- [x] ubenchmark https://github.com/tenstorrent/tt-metal/actions/runs/12129164312
- [x] t3k demo https://github.com/tenstorrent/tt-metal/actions/runs/12131567105

### CODE REVIEW

**Review by pgkeller — COMMENTED** [review]
overall looks good

File: tests/tt_metal/tt_metal/perf_microbenchmark/common/kernels/writer_l1.cpp [hunk 1]
```diff
@@ -200,7 +201,7 @@ FORCE_INLINE void remote_cb_push_back_and_write_pages_(
 
             // issue first write transfer
             while (!noc_cmd_buf_ready(noc, write_cmd_buf));
-            ncrisc_noc_fast_write(
+            ncrisc_noc_fast_write<risc_type, noc_mode>(
```
[the remark this answers is not in the document]
Comment (yugaoTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/common/kernels/writer_l1.cpp:204]: it just do one noc transfer unlike the high-level api, but we are not using this function any where (just as a backup code)

File: tt_metal/hw/firmware/src/brisc.cc [hunk 2]
```diff
@@ -463,6 +466,13 @@ int main() {
 
             wait_ncrisc_trisc();
 
+            if (noc_mode == DM_DYNAMIC_NOC) {
```
[the remark this answers is not in the document]
Comment (yugaoTT) [reply tt_metal/hw/firmware/src/brisc.cc:470]: done

File: tt_metal/hw/inc/dataflow_api.h [hunk 3]
```diff
@@ -27,6 +27,11 @@
 #include "debug/assert.h"
 #include "dev_msgs.h"
 
+#if defined(COMPILE_FOR_BRISC)
+constexpr uint8_t risc_type = static_cast<std::underlying_type_t<TensixProcessorTypes>>(TensixProcessorTypes::DM0);
```
[the remark this answers is not in the document]
Comment (yugaoTT) [reply tt_metal/hw/inc/dataflow_api.h]: done

File: tests/tt_metal/tt_metal/perf_microbenchmark/common/kernels/writer_l1.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (pgkeller) [reply tests/tt_metal/tt_metal/perf_microbenchmark/common/kernels/writer_l1.cpp]: ok.  my only concern is that users look at our code base for examples and I don't want them to use these non-apis

**Review by pgkeller — APPROVED** [review]

**Pull request merged**